File: J9Ch/src/J_8_Data_Structure/Optional_6/_545_Top_k_Largest_Numbers_II_Medium.cpp
/** 545 Top k Largest Numbers II
*/


bool cmp(const int& a, const int& b) {
    return a > b;
}

class Solution {
private:
    priority_queue<int, vector<int>, greater<int>> queue;
    int k;

public:
    Solution(int k) {
        this->k = k;
    }

    void add(int num) {
        if (queue.size() < k) {
            queue.push(num);
        } else if (queue.top() < num) {
            queue.pop();
            queue.push(num);
        }
    }

    vector<int> topk() {
        vector<int> topk;
        int n = queue.size();
        for (int i = 0; i < k && i < n ; ++i) {
            topk.push_back(queue.top());
            queue.pop();
        }

        for (int i = 0; i < n; ++i)
            queue.push(topk[i]);

        sort(topk.begin(), topk.end(), cmp);
        return topk;
    }
};
.  京ICP备16004690号-1

File: J9Ch/src/J_8_DS/Optional_6/python_cpp/_545_Top_k_Largest_Numbers_II_Medium.py
# /** 545 Top k Largest Numbers II


import heapq

class Solution:

    # @param {int} k an integer
    def __init__(self, k):
        self.k = k
        self.nums = []
        heapq.heapify(self.nums)

    # @param {int} num an integer
    def add(self, num):
        if len(self.nums) < self.k:
            heapq.heappush(self.nums, num)
        elif num > self.nums[0]:
            heapq.heappop(self.nums)
            heapq.heappush(self.nums, num)

    # @return {int[]} the top k largest numbers in array
    def topk(self):
        return sorted(self.nums, reverse=True)